### PR TITLE
create MarkedUpTextChunkGenerator

### PR DESCRIPTION
**Pull request by yonadavGit**
## Description
_A brief description of the PR_

## Code Changes
_The following changes were made to the files below_
This pull request introduces a new `linker` module for handling text linking tasks, updates the Celery configuration to include it, and enhances the validation logic for `MarkedUpTextChunk` objects. It also adds a new generator class for creating marked-up text chunks and updates relevant tests to reflect these changes.

### New `linker` module and Celery integration:
* Updated `app.autodiscover_tasks` in `sefaria/celery_setup/app.py` to include the new `sefaria.helper.linker` package.
* Added `sefaria/helper/linker/tasks.py` with two Celery tasks: `hello_world` for basic testing and `link_segment_with_worker` for linking text segments using a linker object. Includes helper functions for processing resolved references and saving marked-up text chunks.

### Marked-up text chunk generation:
* Introduced `MarkedUpTextChunkGenerator` in `sefaria/helper/marked_up_text_chunk_generator.py`, which provides methods for generating and saving marked-up text chunks for text segments across different versions and languages.

### Validation improvements:
* Enhanced `_validate` method in `sefaria/model/marked_up_text_chunk.py` to ensure `spans` cannot be empty and to prevent saving duplicate records with the same primary key but different IDs.

### Test updates:
* Updated tests in `sefaria/model/tests/marked_up_text_chunk.py` to include a new test for empty spans and modified an existing test to use updated object structures.
* Fixed import paths in `sefaria/helper/tests/linker_test.py` and `sefaria/views.py` to reflect the new `linker` module structure. [[1]](diffhunk://#diff-e53978e7b47a370783c254f883cc77bd64d54024f09193a8ef89cafc52e4a0a0L132-R132) [[2]](diffhunk://#diff-e30a05013192fa5147ab7d4c4b7dffb63644ac793c49029bfaa4ab3ee03b5441L355-R355)

## Notes
_Any additional notes go here_

### CODE REVIEW

File: sefaria/helper/linker/tasks.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from sefaria.model import library
+from sefaria.celery_setup.app import app
+from sefaria.model.marked_up_text_chunk import MarkedUpTextChunk
+from sefaria.model.text import Ref
+
+
+
+
+@app.task(name="linker.hello_world")
+def hello_world(words: str):
```
Comment (nsantacruz) [inline sefaria/helper/linker/tasks.py]: remove this task

File: sefaria/helper/linker/tasks.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+@app.task(name="linker.link_segment_with_worker")
+def link_segment_with_worker(raw_input: dict) -> bool:
+    linker = library.get_linker(raw_input["lang"])
+    book_ref = Ref(raw_input["ref"])
+    output = linker.link(raw_input["text"], book_context_ref=book_ref)
+
+    spans = _extract_resolved_spans(output.resolved_refs)
+    if not spans:
+        print(f"No spans found for {raw_input['ref']} in {raw_input['lang']}")
```
Comment (nsantacruz) [inline sefaria/helper/linker/tasks.py]: No reason to print

File: sefaria/helper/linker/tasks.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        return True
```
Comment (nsantacruz) [inline sefaria/helper/linker/tasks.py]: Why does this function return True?

File: sefaria/helper/linker/tasks.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+
+@app.task(name="linker.hello_world")
+def hello_world(words: str):
+    print(words)
+
+
+@app.task(name="linker.link_segment_with_worker")
+def link_segment_with_worker(raw_input: dict) -> bool:
+    linker = library.get_linker(raw_input["lang"])
```
Comment (nsantacruz) [inline sefaria/helper/linker/tasks.py]: create `LinkSegmentInput` class from raw_input

File: sefaria/model/marked_up_text_chunk.py [hunk 5]
```diff
@@ -62,10 +62,13 @@ def _validate(self):
         # Enforce uniqueness
         pkey_query = {k: getattr(self, k) for k in self.pkeys}
 
-        existing = self.load(pkey_query)
-        if existing:
+        existing = self.__class__().load(pkey_query)
+        if existing and existing._id != getattr(self, "_id", None):
             raise DuplicateRecordError(f"{type(self).__name__}._validate(): Duplicate primary key {self.pkeys}, found {pkey_query} to already exist in the database.")
 
+        if not self.spans:
```
Comment (nsantacruz) [inline sefaria/model/marked_up_text_chunk.py]: Move this validation to cerberus schema: ```To enforce that your spans list must contain at least one element, just add a minlength: 1 (or alternatively empty: False) to the list‐rule:```

File: sefaria/helper/marked_up_text_chunk_generator.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+class MarkedUpTextChunkGenerator:
+
+
+    def __init__(self):
+        """Initialize the generator with necessary components."""
+        pass
+
+    def create_and_save_marked_up_text_chunk(self, segment_ref, vtitle, lang, text):
+        linking_args = {
```
Comment (nsantacruz) [inline sefaria/helper/marked_up_text_chunk_generator.py]: Use `LinkSegmentInput` here

File: sefaria/helper/marked_up_text_chunk_generator.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+
+class MarkedUpTextChunkGenerator:
+
+
+    def __init__(self):
+        """Initialize the generator with necessary components."""
+        pass
+
+    def create_and_save_marked_up_text_chunk(self, segment_ref, vtitle, lang, text):
```
Comment (nsantacruz) [inline sefaria/helper/marked_up_text_chunk_generator.py]: use python typing for parameters

File: sefaria/helper/marked_up_text_chunk_generator.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    def _generate_all_versions_for_segment(self, segment_ref: Ref) -> None:
+        for lang, vtitle in self._get_available_versions(segment_ref):
+            text_chunk = TextChunk(segment_ref, lang=lang, vtitle=vtitle)
+            if not text_chunk.text:
+                logger.debug(f"No text found for {segment_ref.normal()}, {vtitle}, {lang}")
+                continue
+            self.generate(segment_ref, lang, vtitle)
+
+    def generate(self, ref: Ref, lang: str, vtitle: str) -> None:
```
Comment (nsantacruz) [inline sefaria/helper/marked_up_text_chunk_generator.py]: move public functions to the top of class

File: sefaria/helper/marked_up_text_chunk_generator.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        except Exception as e:
+            logger.error(f"Failed to create/save MarkedUpTextChunk for {segment_ref.normal()}, {vtitle}, {lang}: {e}")
+            raise
+
+
+    def _get_available_versions(self, segment_ref: Ref) -> List[Tuple[str, str]]:
+        versions = []
+
+        # Get the index for this ref
+        index = segment_ref.index
```
Comment (nsantacruz) [inline sefaria/helper/marked_up_text_chunk_generator.py]: look into using `Ref.versionset()` instead. If this works, consider removing this function entirely!

File: sefaria/helper/linker/tasks.py [hunk 10]
```diff
@@ -7,23 +7,37 @@
...
+from dataclasses import dataclass, asdict
 
+@dataclass(frozen=True)
+class LinkingArgs:
+    ref: str
+    text: str
+    lang: str
+    vtitle: str
 
+    @staticmethod
```
Comment (nsantacruz) [inline sefaria/helper/linker/tasks.py]: these two functions add almost no benefit since they are so simple. I would remove them.

File: sefaria/helper/linker/tasks.py [hunk 11]
```diff
@@ -19,8 +19,7 @@ def link_segment_with_worker(raw_input: dict) -> bool:
 
     spans = _extract_resolved_spans(output.resolved_refs)
     if not spans:
-        print(f"No spans found for {raw_input['ref']} in {raw_input['lang']}")
-        return True
+        return
```
Comment (nsantacruz) [inline sefaria/helper/linker/tasks.py:30]: remove superfluous return

**Review by nsantacruz — APPROVED** [review]
